dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/Dependency.java

```diff
@@ -0,0 +1,229 @@
+package com.dat3m.dartagnan.program.analysis;
+
+import com.dat3m.dartagnan.program.Program;
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.Thread;
+import com.dat3m.dartagnan.program.event.core.CondJump;
+import com.dat3m.dartagnan.program.event.core.Event;
+import com.dat3m.dartagnan.program.event.core.MemEvent;
+import com.dat3m.dartagnan.program.event.core.utils.RegReaderData;
+import com.dat3m.dartagnan.program.event.core.utils.RegWriter;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Verify.verify;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.IntStream.range;
+
+/**
+ * Computes direct providers for register values, based on when a register is used.
+ */
+public final class Dependency {
+
```

Comment:
The key in `map` and the list elements in `may` and `must` are always `RegReader` and `RegWriter`. Why, instead of using `Event`, you don't use the more precise type? This allows to simplify parts of the code like the assertion [here](https://github.com/hernanponcedeleon/Dat3M/blob/961a16415ffbec4b26882c6830c20ae36ef42ef0/dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java#L249) or the cast [here](https://github.com/hernanponcedeleon/Dat3M/blob/961a16415ffbec4b26882c6830c20ae36ef42ef0/dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java#L258) (there are probably more).

Reply:
There is no `RegReader` interface, but `RegReaderData` and `MemEvent`, and I see no benefit in introducing such a new interface.
`RegWriter` does not extend the interface of `Event`, and thus had to be cast to `Event` when needed, like `writer.exec()` or `new Tuple(writer,reader)`.  This was one of the few points of #202.